lightautoml/ml_algo/torch_based/nn_models.py

```diff
@@ -975,3 +992,112 @@ def forward(self, embedded: torch.Tensor) -> torch.Tensor:
+            n_out: Output dimension, 1 for binary prediction.
+            embedding_size: Embeddings size.
+            depth: Number of Attention Blocks inside Transformer.
+            heads: Number of heads in Attention.
+            attn_dropout: Post-Attention dropout.
+            ff_dropout: Feed-Forward Dropout.
+            dim_head: Attention head dimension.
+            return_attn: Return attention scores or not.
+            num_enc_layers: Number of Transformer layers.
+            device: Device to compute on.
+    """
+
+    def __init__(
+        self,
+        *,
+        pooling: str = "concat",
+        n_out: int = 1,
+        embedding_size: int = 16,
+        depth: int = 2,
+        heads: int = 1,
+        attn_dropout: float = 0.1,
+        ff_dropout: float = 0.1,
+        dim_head: int = 64,
+        return_attn: bool = False,
+        num_enc_layers: int = 2,
+        device: Union[str, torch.device] = "cuda:0",
+        **kwargs,
+    ):
+        super(FTTransformer, self).__init__()
+        self.return_attn = return_attn
+        # self.num_enc_layers = num_enc_layers
+        self.device = device
+        if pooling == "cls":
+            self.pooling = SequenceClsPooler()
+        elif pooling == "mean":
+            self.pooling = SequenceAvgPooler()
+        elif pooling == "sum":
+            self.pooling = SequenceSumPooler()
+        elif pooling == "concat":
+            self.pooling = SequenceConcatPooler()
+        elif pooling == "none":
+            self.pooling = SequenceIndentityPooler()
+
+        # transformer
+
+        self.transformer = nn.Sequential(
+            *nn.ModuleList(
+                [
+                    Transformer(
+                        dim=embedding_size,
+                        depth=depth,
+                        heads=heads,
+                        dim_head=dim_head,
+                        attn_dropout=attn_dropout,
+                        ff_dropout=ff_dropout,
+                        return_attn=self.return_attn,
+                    )
+                    for _ in range(num_enc_layers)
+                ]
+            )
+        )
+
+        # to logits
+        if pooling == "concat":
+            self.to_logits = nn.Sequential(nn.BatchNorm1d(embedding_size * 2), nn.Linear(embedding_size * 2, n_out))
+        else:
+            self.to_logits = nn.Sequential(nn.BatchNorm1d(embedding_size), nn.Linear(embedding_size, n_out))
+
+        self.cls_token = nn.Embedding(2, embedding_size)
+
+    def forward(self, embedded):
+        """Transform the input tensor.
+
+        Args:
+            embedded : torch.Tensor
+                embedded fields
+
+        Returns:
+            torch.Tensor
+
+        """
+        cls_token = torch.unsqueeze(
+            self.cls_token(torch.ones(embedded.shape[0], dtype=torch.int).to(self.device)), dim=1
+        )
+        x = torch.cat((cls_token, embedded), dim=1)
+
+        if not self.return_attn:
```

Comment:
Try to avoid unnecessary if in forward. You can always return attn.

Reply:
done